### PR TITLE
Fix #2: Check node id uniqueness on join

### LINKED ISSUE #2: Adding a node with the same id as a previously existing node should be forbidden
After creating a cluster with 2 nodes,
then running "RAFT.NODE remove 2"
Then running another server (with new log) and uid 2, when trying to run "RAFT.CLUSTER JOIN localhost:5001" the new server crashes with error:
"Assertion failed: (node), function raft_handle_append_cfg_change, file src/raft_server.c, line 1227."

### PR DESCRIPTION
**Pull request by gavrie**
- We keep a list of all node ids that were ever members of the cluster
- When adding a node, this list is checked
- If the node id has ever been used, the node will not be joined

### CODE REVIEW

File: tests/integration/test_membership.py [hunk 1]
```diff
@@ -135,6 +140,29 @@ def test_full_cluster_remove(cluster):
         assert cluster.node(node_id).raft_info()['state'] == 'uninitialized'
 
 
+def test_remove_and_rejoin_node_with_same_id_fails(cluster):
```
Comment (yossigo) [inline tests/integration/test_membership.py:143]: Full coverage here should consider the different persistence states, i.e. make sure we remember used Ids also if loaded from snapshot, restarted from log, etc.

File: raft.c [hunk 2]
```diff
@@ -934,7 +943,7 @@ static raft_node_id_t makeRandomNodeId(RedisRaftCtx *rr)
     do {
         RedisModule_GetRandomBytes((unsigned char *) &id, sizeof(id));
         id &= ~(1<<31);
-    } while (!id || (rr->raft && raft_get_node(rr->raft, id) != NULL));
+    } while (!id || (rr->raft && raft_get_node(rr->raft, id) != NULL) || hasNodeIdBeenUsed(rr, id));
```
Comment (yossigo) [inline raft.c:946]: Am I missing something here? `makeRandomNodeId()` would always get called by a node that is not yet part of the cluster, so `hasNodeIdBeenUsed()` should always be `false`.

File: node.c [hunk 3]
```diff
@@ -329,14 +329,20 @@ void handleNodeAddResponse(redisAsyncContext *c, void *r, void *privdata)
...
+
+        LOG_INFO("Joined Raft cluster, node id: %u, dbid: %.*s\n",
+                node_id,
                 reply->element[1]->len, reply->element[1]->str);
 
         strncpy(rr->snapshot_info.dbid, reply->element[1]->str, reply->element[1]->len);
         rr->snapshot_info.dbid[RAFT_DBID_LEN] = '\0';
 
-        rr->config->id = reply->element[0]->integer;
+        rr->config->id = node_id;
```
[the remark this answers is not in the document]
Comment (gavrie) [reply node.c:341]: Sure, I agree -- my changes didn't modify this assumption. I just added a local node_id variable to improve readability.

File: node.c [hunk 4]
```diff
@@ -329,14 +329,20 @@ void handleNodeAddResponse(redisAsyncContext *c, void *r, void *privdata)
...
+
+        /* TODO: Remove log message */
+        LOG_INFO("handleNodeAddResponse: addUsedNodeId %d\n", node_id);
+        addUsedNodeId(rr, node_id);
```
[the remark this answers is not in the document]
Comment (gavrie) [reply node.c:345]: Wasn't aware of it, sounds good!

File: raft.c (hunk 2 above)
Comment (gavrie) [reply raft.c:946]: Well, theoretically there could be a clash where the random id is equal to a previously used id, which we check for.
In practice the chance is low, but why not check for it?

File: tests/integration/test_membership.py (hunk 1 above)
Comment (gavrie) [reply tests/integration/test_membership.py:143]: 👍

File: raft.c (hunk 2 above)
Comment (gavrie) [reply raft.c:946]: Per out last discussion, we'll keep this.

**Review by yossigo — APPROVED** [review]

**Review by yossigo — APPROVED** [review]

**Pull request merged**